File: interfaces/examples/C/basic_cg.c
/*
 * basic_cg.c — minimal example: solve a 5x5 SPD system with CG.
 *
 * A = tridiag(-1, 2, -1),  b = [1, 0, 0, 0, 1]^T
 *
 * Compile (after building libkrylov with juliac — see interfaces/README.md):
 *
 *   gcc -o basic_cg interfaces/examples/C/basic_cg.c \
 *       -I interfaces/build/include \
 *       interfaces/build/lib/libkrylov.so \
 *       -Wl,-rpath,'$ORIGIN/../lib/julia'
 *
 * Expected output:
 *   Solved: yes   niter: 3   time: ...
 *   x = [ 1.00  1.00  1.00  1.00  1.00 ]
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "krylov.h"

/* -------------------------------------------------------------------------
 * Problem data
 * ------------------------------------------------------------------------- */

#define N 5

/* Tridiagonal matrix stored as three diagonals for simplicity. */
typedef struct {
  int    n;
  double diag[N];   /* main diagonal */
  double off[N-1];  /* sub/super diagonal */
} TriDiag;

/* Matvec callback:  y = A * x  */
static void matvec_A(const void *xv, void *yv, void *userdata)
{
  const double *x  = (const double *)xv;
  double       *y  = (double *)yv;
  const TriDiag *A = (const TriDiag *)userdata;
  int n = A->n;

  for (int i = 0; i < n; i++) {
    y[i] = A->diag[i] * x[i];
    if (i > 0)   y[i] += A->off[i-1] * x[i-1];
    if (i < n-1) y[i] += A->off[i]   * x[i+1];
  }
}

/* -------------------------------------------------------------------------
 * main
 * ------------------------------------------------------------------------- */

int main(void)
{
  /* Build A = tridiag(-1, 2, -1) */
  TriDiag A;
  A.n = N;
  for (int i = 0; i < N;   i++) A.diag[i] = 2.0;
  for (int i = 0; i < N-1; i++) A.off[i]  = -1.0;

  /* Right-hand side */
  double b[N] = {1.0, 0.0, 0.0, 0.0, 1.0};

  /* Solution buffer */
  double x[N];

  /* -----------------------------------------------------------------------
   * Create workspace for CG, double precision, CPU
   * --------------------------------------------------------------------- */
  void *ws = NULL;
  int ret = krylov_workspace_create(KRYLOV_CG, N, N,
                                    KRYLOV_FLOAT64, KRYLOV_CPU,
                                    NULL,    /* workspace options (NULL = defaults) */
                                    &ws);
  if (ret != 0) {
    fprintf(stderr, "krylov_workspace_create failed (%d)\n", ret);
    return 1;
  }

  /* -----------------------------------------------------------------------
   * Solve
   * --------------------------------------------------------------------- */
  KrylovOptions opts = krylov_default_options();
  opts.atol = 1e-10;
  opts.rtol = 1e-10;

  ret = krylov_solve(ws,
                     matvec_A,   /* y = A*x */
                     NULL,       /* y = A'*x  (CG doesn't need it) */
                     NULL,       /* no left preconditioner */
                     NULL,       /* no right preconditioner */
                     b,          /* right-hand side b (size m) */
                     NULL,       /* c = NULL  (CG only needs one RHS) */
                     &A,         /* userdata forwarded to matvec_A */
                     &opts);     /* solver options (NULL = all defaults) */
  if (ret != 0) {
    fprintf(stderr, "krylov_solve failed (%d)\n", ret);
    krylov_workspace_free(ws);
    return 1;
  }

  /* -----------------------------------------------------------------------
   * Retrieve results
   * --------------------------------------------------------------------- */
  ret = krylov_get_x(ws, x, N);
  if (ret != 0) {
    fprintf(stderr, "krylov_get_x failed (%d)\n", ret);
    krylov_workspace_free(ws);
    return 1;
  }

  printf("Solved: %s   niter: %d   time: %.3e s\n",
         krylov_is_solved(ws) ? "yes" : "no",
         krylov_niter(ws),
         krylov_elapsed_time(ws));

  printf("x = [");
  for (int i = 0; i < N; i++)
    printf(" %.2f", x[i]);
  printf(" ]\n");

  /* -----------------------------------------------------------------------
   * Free workspace
   * --------------------------------------------------------------------- */
  krylov_workspace_free(ws);

  return 0;
}

File: interfaces/examples/Fortran/basic_cg.f90
! basic_cg.f90 — solve a 5x5 SPD system with CG.
!
! A = tridiag(-1, 2, -1),  b = [1, 0, 0, 0, 1]^T
!
! Compile:
!   gfortran -o basic_cg interfaces/examples/Fortran/basic_cg.f90 \
!       interfaces/build/lib/libkrylov.so

program basic_cg
  use iso_c_binding
  implicit none
  include '../../include/krylov.f90'

  integer, parameter :: n = 5

  ! Problem data — passed to the callback as userdata
  real(c_double), target :: diag(n), off(n-1)

  real(c_double), target  :: b(n), x(n)
  type(c_ptr)             :: ws
  integer(c_int)          :: ret

  ! -------------------------------------------------------------------------
  ! Build A = tridiag(-1, 2, -1)
  ! -------------------------------------------------------------------------
  diag = 2.0_c_double
  off  = -1.0_c_double

  ! -------------------------------------------------------------------------
  ! Right-hand side
  ! -------------------------------------------------------------------------
  b = 0.0_c_double
  b(1) = 1.0_c_double
  b(n) = 1.0_c_double

  ! -------------------------------------------------------------------------
  ! Create workspace
  ! -------------------------------------------------------------------------
  ret = krylov_workspace_create(KRYLOV_CG, n, n,            &
                                KRYLOV_FLOAT64, KRYLOV_CPU, &
                                c_null_ptr,                 &  ! workspace options (defaults)
                                ws)
  if (ret /= 0) then
    write(*,*) "krylov_workspace_create failed:", ret
    stop 1
  end if

  ! -------------------------------------------------------------------------
  ! Solve
  ! -------------------------------------------------------------------------
  block
    type(KrylovOptions), target :: opts
    opts = krylov_default_options()
    opts%atol = 1.0d-10
    opts%rtol = 1.0d-10

    ret = krylov_solve(ws,                  &
                       c_funloc(matvec_A),  &  ! y = A*x
                       c_null_funptr,       &  ! y = A'*x  (CG doesn't need it)
                       c_null_funptr,       &  ! no left preconditioner
                       c_null_funptr,       &  ! no right preconditioner
                       c_loc(b),            &  ! right-hand side b (size m)
                       c_null_ptr,          &  ! c = NULL  (CG only needs one RHS)
                       c_loc(diag),         &  ! userdata: diagonal array
                       c_loc(opts))            ! solver options
  end block
  if (ret /= 0) then
    write(*,*) "krylov_solve failed:", ret
    ret = krylov_workspace_free(ws)
    stop 1
  end if

  ! -------------------------------------------------------------------------
  ! Retrieve solution
  ! -------------------------------------------------------------------------
  ret = krylov_get_x(ws, c_loc(x), int(n, c_int))
  if (ret /= 0) then
    write(*,*) "krylov_get_x failed:", ret
    ret = krylov_workspace_free(ws)
    stop 1
  end if

  ! -------------------------------------------------------------------------
  ! Print results
  ! -------------------------------------------------------------------------
  write(*,'(A,L1,A,I0,A,ES10.3,A)') &
    "Solved: ", (krylov_is_solved(ws) == 1), &
    "   niter: ", krylov_niter(ws),          &
    "   time: ", krylov_elapsed_time(ws), " s"
  write(*,'(A)', advance='no') "x = ["
  write(*,'(5F6.2)', advance='no') x
  write(*,'(A)') " ]"

  ! -------------------------------------------------------------------------
  ! Free workspace
  ! -------------------------------------------------------------------------
  ret = krylov_workspace_free(ws)

contains

  ! -------------------------------------------------------------------------
  ! Matvec callback:  y = A*x   (A = tridiag(-1, 2, -1))
  ! -------------------------------------------------------------------------
  subroutine matvec_A(x_ptr, y_ptr, userdata) bind(c)
    use iso_c_binding
    type(c_ptr), value :: x_ptr, y_ptr, userdata

    real(c_double), pointer :: xv(:), yv(:), dg(:)
    integer :: i

    call c_f_pointer(x_ptr,    xv, [n])
    call c_f_pointer(y_ptr,    yv, [n])
    call c_f_pointer(userdata, dg, [n])   ! diag array

    do i = 1, n
      yv(i) = dg(i) * xv(i)
      if (i > 1) yv(i) = yv(i) - xv(i-1)
      if (i < n) yv(i) = yv(i) - xv(i+1)
    end do
  end subroutine matvec_A

end program basic_cg
